include/SDL3/SDL_pen.h

```diff
@@ -0,0 +1,265 @@
+ *
+ * \param instance_id The pen to query.
+ *
+ * \returns The corresponding pen GUID; persistent across multiple sessions.
+ *     If "instance_id" is ::SDL_PEN_INVALID, returns an all-zeroes GUID.
+ *
+ * \since This function is available since SDL 3.TBD
+ *
+ * \sa SDL_PenForID()
+ */
+extern DECLSPEC SDL_GUID SDLCALL SDL_GetPenGUID(SDL_PenID instance_id);
+
+/**
+ * Checks whether a pen is still attached.
+ *
+ * If a pen is detached, it will not show up for ::SDL_GetPens().
+ * Other operations will still be available but may return default values.
+ *
+ * \param instance_id A pen ID.
+ * \returns SDL_TRUE if "instance_id" is valid and the corresponding pen is attached, or
+ *     SDL_FALSE otherwise.
+ *
+ * \since This function is available since SDL 3.TBD
+ */
+extern DECLSPEC SDL_bool SDLCALL SDL_PenConnected(SDL_PenID instance_id);
+
+/**
+ * Retrieves a human-readable description for a ::SDL_PenID.
+ *
+ * \param instance_id The pen to query.
+ *
+ * \returns A string that contains the name of the pen, intended for human consumption.
+ *     The string might or might not be localised, depending on platform settings.
+ *     It is not guaranteed to be unique; use ::SDL_GetPenGUID() for (best-effort)
+ *     unique identifiers.
+ *     The pointer is managed by the SDL pen subsystem and must not be deallocated.
+ *     The pointer remains valid until SDL is shut down.
+ *     Returns NULL on error (cf. ::SDL_GetError())
+ *
+ * \since This function is available since SDL 3.TBD
+ */
+extern DECLSPEC const char *SDLCALL SDL_GetPenName(SDL_PenID instance_id);
+
+/**
+ * Pen capabilities, as reported by ::SDL_GetPenCapabilities()
+ */
+typedef struct SDL_PenCapabilityInfo
+{
+    float max_tilt;    /**< Physical maximum tilt values, or SDL_PEN_INFO_UNKNOWN .  Pens can't typically tilt all the way to 90 degrees, so this value is usually less than 1.0. */
+    Uint32 wacom_id;   /**< For Wacom devices: wacom tool type ID, otherwise 0 (useful e.g. with libwacom) */
+    Sint8 num_buttons; /**< Number of pen buttons (not counting the pen tip), or SDL_PEN_INFO_UNKNOWN */
+} SDL_PenCapabilityInfo;
+
+/**
+ * Retrieves capability flags for a given ::SDL_PenID.
+ *
+ * \param instance_id The pen to query.
+ * \param[out] capabilities Detail information about pen capabilities, such as the number of buttons
+ *
+ * \returns a set of capability flags, cf. \link SDL_PEN_CAPABILITIES \endlink.  Returns 0 on error
+ *     (cf. ::SDL_GetError())
+ *
+ * \since This function is available since SDL 3.TBD
+ */
+extern DECLSPEC Uint32 SDLCALL SDL_GetPenCapabilities(SDL_PenID instance_id, SDL_PenCapabilityInfo *capabilities);
+
+/**
+ * Retrieves the pen type for a given ::SDL_PenID.
+ *
+ * \param instance_id The pen to query.
+ * \returns The corresponding pen type (cf. ::SDL_PenSubtype) or 0 on error.
+ * \since This function is available since SDL 3.TBD
+ */
+extern DECLSPEC SDL_PenSubtype SDLCALL SDL_GetPenType(SDL_PenID instance_id);
```

Comment:
I wonder if this will be too inflexible and cumbersome to need to create a new pen each time the interaction changes between a pen tip and eraser.

The way existing pen APIs seem to handle this is by having a separate "tool type" parameter or flag associated with each active pen interaction. The tool type would be a property of the specific pen interaction rather than a property of the pen itself.

See:
https://developer.android.com/reference/android/view/MotionEvent#getToolType(int)
https://learn.microsoft.com/en-us/windows/win32/inputmsg/pen-flags-constants
https://docs.kernel.org/input/event-codes.html?highlight=evdev#ev-key

Reply:
That's basically all that I really noticed. I do agree with your comment about having separate pen down/up events rather than using a button flag for that.